### ISSUE TITLE
Bash history incomplete/missing recent entries

### ISSUE THREAD

**Issue by YoungElPaso** [MicrosoftDocs/vsonline#503]
I noticed that when the environment suspends and I subsequently resume a good chunk of my bash history file is missing.  This has happened at least twice and I've confirmed by backing up history to a text file before suspending.

By itself, this isn't a huge deal although I'm sure a lot of people use history to keep track of things.

But what is worrying is I wonder what other things, (maybe env variables, aliases etc), that I might use in bash or elsewhere are also getting 'lost'.  

To me, a big strength of using a VS Codespace to work on a given project from anywhere would be resuming the previous state of that work - bash history and all.  Similar to what I experience locally but available anywhere with a browser.  

TLDR: bash history (and maybe other stuff?) not persisting across sessions.

**Comment by jkeech** [comment]
@YoungElPaso, do you have some specific repro steps? Is it containing older entries and not newer ones, or perhaps it's the older entries that are getting truncated?

The default `.bashrc` config has the following set (unless you are changing them with dotfiles or something else):

```bash
HISTSIZE=1000
HISTFILESIZE=2000
```

Is it possible that you have more lines than that in your history? What does `wc -l ~/.bash_history ` return?

**Comment by jkeech** [comment]
It looks like when the container shuts down without the interactive shells exiting cleanly first, that any history currently stored in memory might not be written to the history file

**Comment by jkeech** [comment]
I confirmed that if you have some pending bash history in memory and the environment suspends without you leaving gracefully, then the bash process doesn't cleanly exit and get a chance to write the history in memory to the `~/.bash_history` file. Need to look into ways to make this work better out of the box.

**Comment by jkeech** [comment]
After some more experimentation, the best way for us to ensure the bash history gets written it to first kill the VS Code server process inside the container before executing `docker stop` during suspension. Killing the VS Code server process will tear down the child bash processes correctly which gives them a chance to write out the history to disk before the container is stopped.

**Comment by jkeech** [comment]
The fix for this has been merged and will be deploying this week

**Comment by dmgardiner25** [comment]
The changes were deployed and can be seen from codespaces run on desktop but we've run into an issue that is preventing the changes from being seen on the web. Since that's a separate issue, I'm going to go ahead and close this issue.
